### ISSUE TITLE
Bootstrap issues with musl/armhf

### ISSUE THREAD

**Issue by philhofer** [gittup/tup#323]
I'm trying to build tup from source on Alpine Linux, which uses musl-libc, so naturally there are some teething issues to sort out.

The first issue I encountered was that there are a few `read` calls in `src/tup/server/master_fork.c` that don't handle `EINTR`, which means that if a child crashes and SIGCHLD is received during a call to `read`, you get an unhelpful `read: Operation interrupted` error message. This fix is easy:
```
diff --git a/src/tup/server/master_fork.c b/src/tup/server/master_fork.c
index 4ade0bda..0d0d8c01 100644
--- a/src/tup/server/master_fork.c
+++ b/src/tup/server/master_fork.c
@@ -195,8 +195,10 @@ int server_pre_init(void)
                perror("close(msd[0])");
                return -1;
        }
+read_rc_again:
        rc = read(msd[1], &c, 1);
        if(rc < 0) {
+               if(errno == EINTR) goto read_rc_again;
                perror("read");
                return -1;
        }
@@ -310,6 +312,7 @@ static int read_all_internal(int sd, void *dest, int size, int line)
        while(bytes_read < size) {
                rc = read(sd, p + bytes_read, size - bytes_read);
                if(rc < 0) {
+                       if(errno == EINTR) continue;
                        perror("read");
                        fprintf(stderr, "tup error: Unable to read from the master fork socket (read_all called from line %i).\n", line);
                        return -1;

```

The second issue I haven't yet been able to solve. The `chdir` into the job subdirectory fails with `EINVAL` for reasons I don't understand.
```
...
LOOKUP /@tupjob-1364/home/phil/tup/src/lua
getattr /@tupjob-1364/home/phil/tup/src/lua
   NODEID: 23
   unique: 25, success, outsize: 144
unique: 26, opcode: LOOKUP (1), nodeid: 22, insize: 44, pid: 6250
LOOKUP /@tupjob-1362/home/phil/tup/src
getattr /@tupjob-1362/home/phil/tup/src
   NODEID: 24
   unique: 26, success, outsize: 144
unique: 27, opcode: ACCESS (34), nodeid: 23, insize: 48, pid: 6249
access /@tupjob-1364/home/phil/tup/src/lua 01
   unique: 27, error: -22 (Invalid argument), outsize: 16
unique: 28, opcode: LOOKUP (1), nodeid: 24, insize: 44, pid: 6250
LOOKUP /@tupjob-1362/home/phil/tup/src/lua
getattr /@tupjob-1362/home/phil/tup/src/lua
   NODEID: 25
   unique: 28, success, outsize: 144
unique: 29, opcode: ACCESS (34), nodeid: 25, insize: 48, pid: 6250
access /@tupjob-1362/home/phil/tup/src/lua 01
   unique: 29, error: -22 (Invalid argument), outsize: 16
*  1) src/lua: CC ldebug.c
DEBG: chdir(home/phil/tup/src/lua)
chdir: Invalid argument
tup error: Unable to chdir to 'home/phil/tup/src/lua'
 *** tup errors ***
 *** Command ID=1366 failed with return value 1
tup error: Expected to write to file 'ldebug.o' from cmd 1366 but didn't
```

I'll send an update if I manage to sort this out myself, but I would appreciate a hint or two.

Thanks,
Phil

**Comment by philhofer** [comment]
Update: `faccessat(AT_SYMLINK_NOFOLLOW)` appears to be unsupported:
http://www.openwall.com/lists/musl/2015/02/05/2

I'll send a patch to disable `AT_SYMLINK_NOFOLLOW` when musl is detected.

**Comment by philhofer** [comment]
Closing this issue b/c it was fixed in PR #339 more than a year ago.
